### ISSUE TITLE
haproxy 2.4 memory leak

### ISSUE THREAD

**Issue by danielleontiev** [haproxy/haproxy#1499]
### Detailed Description of the Problem

haproxy 2.4 does not free memory if a lot of HTTP 500 errors occurring in the backend. In contrast, haproxy 2.3 behaviour looks valid.

Below metrics for both haproxy 2.3 and 2.4 will be shown for comparison.

Consider the following workload:

haproxy 2.3:

![image](https://user-images.githubusercontent.com/24596613/147246924-4886c8a8-d13d-4aac-b2c2-01a1af21d223.png)

haproxy 2.4

![image](https://user-images.githubusercontent.com/24596613/147246957-84a63038-d373-4b92-903c-9a3b0cf43aa4.png)

There are ~10% of `200 OK` responses from backend and ~90% `500 Internal Server Error` responses. This is provided by this example application:

```go
package main

import (
	"math/rand"
	"net/http"
)

func main() {
	http.HandleFunc("/sample", sampleHandler)
	http.HandleFunc("/health", func(rw http.ResponseWriter, r *http.Request) { rw.WriteHeader(200) })

	if err := http.ListenAndServe(":9000", nil); err != nil {
		panic(err)
	}
}

func sampleHandler(rw http.ResponseWriter, r *http.Request) {
	if rand.Float32() > 0.1 {
		rw.WriteHeader(http.StatusInternalServerError)
		return
	}

	rw.WriteHeader(http.StatusOK)
}
```

Memory usage for haproxy 2.3 and haproxy 2.4:

![image](https://user-images.githubusercontent.com/24596613/147247244-33e9a582-74c0-4407-9300-e0014d67eaf9.png)

It seems that haproxy 2.4 does not free memory in comparison with haproxy 2.3

The actual triggering point for the leak is `HTTP 500` errors from the backend. For example, if the application is changed as follows (now it will return only `200 OK`):

```go
package main

import (
	"net/http"
)

func main() {
	http.HandleFunc("/sample", sampleHandler)
	http.HandleFunc("/health", func(rw http.ResponseWriter, r *http.Request) { rw.WriteHeader(200) })

	if err := http.ListenAndServe(":9000", nil); err != nil {
		panic(err)
	}
}

func sampleHandler(rw http.ResponseWriter, r *http.Request) {
	rw.WriteHeader(http.StatusOK)
}
```

Than memory leak does not appear anymore. Workload:

haproxy 2.3:

![image](https://user-images.githubusercontent.com/24596613/147247614-b46bf993-30fa-45ad-b235-6e3de05cb669.png)

haproxy 2.4:

![image](https://user-images.githubusercontent.com/24596613/147247650-7f2711e0-709e-4d19-b75d-3f6ed4e8eda5.png)

Memory usage:

![image](https://user-images.githubusercontent.com/24596613/147247679-fa7a3586-9a8a-4e84-9105-8f549614ef91.png)

### Expected Behavior

The leak should not appear

### Steps to Reproduce the Behavior

Clone the repository https://github.com/danielleontiev/haproxy-memory-leak and go through steps in `README.md`

### What is your configuration?

```haproxy
https://github.com/danielleontiev/haproxy-memory-leak/tree/main/haproxy
```

### Output of `haproxy -vv`

```plain
**haproxy 2.3**

HA-Proxy version 2.3.10-4764f0e 2021/04/23 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2022.
Known bugs: http://www.haproxy.org/bugs/bugs-2.3.10.html
Running on: Linux 5.10.76-linuxkit #1 SMP Mon Nov 8 10:21:19 UTC 2021 x86_64
Build options :
  TARGET  = linux-musl
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1
  DEBUG   = 

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED -BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -CLOSEFROM +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=6).
Built with OpenSSL version : OpenSSL 1.1.1k  25 Mar 2021
Running on OpenSSL version : OpenSSL 1.1.1k  25 Mar 2021
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.6
Built with network namespace support.
Built with the Prometheus exporter as a service
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.36 2020-12-04
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 10.3.1 20210424

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2
            fcgi : mode=HTTP       side=BE        mux=FCGI
       <default> : mode=HTTP       side=FE|BE     mux=H1
       <default> : mode=TCP        side=FE|BE     mux=PASS

Available services : prometheus-exporter
Available filters :
	[SPOE] spoe
	[CACHE] cache
	[FCGI] fcgi-app
	[COMP] compression
	[TRACE] trace




**haproxy 2.4**

HAProxy version 2.4.9-f8dcd9f 2021/11/24 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2026.
Known bugs: http://www.haproxy.org/bugs/bugs-2.4.9.html
Running on: Linux 5.10.76-linuxkit #1 SMP Mon Nov 8 10:21:19 UTC 2021 x86_64
Build options :
  TARGET  = linux-musl
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_PROMEX=1
  DEBUG   = 

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED -BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -CLOSEFROM -ZLIB +SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS -OT -QUIC +PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=6).
Built with OpenSSL version : OpenSSL 1.1.1l  24 Aug 2021
Running on OpenSSL version : OpenSSL 1.1.1l  24 Aug 2021
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.6
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.39 2021-10-29
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 10.3.1 20211027

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG

Available services : prometheus-exporter
Available filters :
	[SPOE] spoe
	[CACHE] cache
	[FCGI] fcgi-app
	[COMP] compression
	[TRACE] trace
```

**Comment by wtarreau** [comment]
Thanks for your detailed report. We're currently saturated as less developers are available and other issues are already under investigation. I'm copy-pasting your config here for reference to ease analysis for everyone:
```
global
  log stdout format raw local0 debug
  stats socket ipv4@*:9999 level admin
  stats timeout 2m
  maxconn 14000
  tune.bufsize 128000
  nbproc 1
  nbthread 4
  cpu-map auto:1/1-4 0-3

resolvers dnsserver
  parse-resolv-conf
  resolve_retries 3
  timeout resolve 5000ms
  timeout retry 5000ms
  hold valid 10s
  hold refused 7000ms
  hold nx 7000ms
  hold timeout 7000ms
  hold other 7000ms
  hold obsolete 7000ms

defaults
  mode http
  default-server init-addr last,libc,none
  option log-health-checks
  timeout connect 5000ms
  timeout server 60000ms
  option  dontlognull
  log global
  log-format "%ci:%cp [%tr] %ft %b/%s %TR/%Tw/%Tc/%Tr/%Ta %ST %B %CC %CS %tsc %ac/%fc/%bc/%sc/%rc %sq/%bq %hr %hs %{+Q}r"

frontend stats
  bind *:8080
  option http-use-htx
  http-request use-service prometheus-exporter if { path /metrics }
  stats enable
  stats uri /stats
  stats refresh 10s

frontend proxy-frontend
  bind *:9000
  timeout client 60000ms
  use_backend application

backend application
  mode http
  balance roundrobin
  default-server fall 2 rise 1
  retry-on all-retryable-errors
  retries 1
  option redispatch
  server us-weaver-primary 127.0.0.1:5008 resolvers dnsserver track check_application/checker

listen application-proxy
  bind 127.0.0.1:5008
  http-request set-header Host application
  server application application:9000 resolvers dnsserver

backend check_application
  option httpchk
  http-check send hdr host application meth GET uri /health
  http-check expect status 200
  default-server fall 2 rise 1
  server checker application:9000 resolvers dnsserver check inter 3s
```
I'm seeing a "retry-on all-retryable-errors", which seriously looks like a good candidate: in order to handle request retries, we have to duplicate the pending request into a dedicated buffer. If for any reason an error path manages to reset this buffer without freeing it, it could definitely cause such a leak. If it's easy for you I would appreciate it if you could try to comment it out and re-test.

**Comment by danielleontiev** [comment]
Yes, you are right. Commenting out `retry-on all-retryable-errors` helped.

Here is workload:

![image](https://user-images.githubusercontent.com/24596613/147276457-c74c1170-b494-4b5b-886c-f521313d5aee.png)

And memory usage:

![image](https://user-images.githubusercontent.com/24596613/147276484-d075bef8-8507-4aee-82b4-5c4329e3463e.png)

**Comment by wtarreau** [comment]
Cool, many thanks! So now we know better where to look for the problem. I don't hope to give you any news before next week however.

**Comment by danielleontiev** [comment]
Thank you! Looking forward for the news

**Comment by wtarreau** [comment]
OK I could now reproduce it with a minimal self-standing config:
```
global
  stats socket /tmp/sock1 level admin
  stats timeout 2m

defaults
  mode http
  timeout client 60000ms
  timeout connect 5000ms
  timeout server 60000ms

listen proxy
  bind *:9000
  retry-on all-retryable-errors
  retries 1
  option redispatch
  server app 127.0.0.1:5008

frontend dummy-app
  bind 127.0.0.1:5008
  http-request return status 500
```

With this I send requests to port 9000 and I'm seeing on "show pools" that the `sockaddr` entries are continuously growing:

```
$ socat - /tmp/sock1 <<< "show pools" | sort -nk7
...
  - Pool sockaddr (128 bytes) : 615955 allocated (78842240 bytes), 615955 used, needed_avg 615950, 0 failures, 1 users, @0x23bbe40 [SHARED]
Total: 50 pools, 81519456 bytes allocated, 81519456 used.
```

The good news is that it's doing this as well in 2.6-dev. I suspect this has soemthing to do with alloc_bind_address() that is unconditionally called in connect_server(). Maybe there's a code path where it's not released. The addresses allocated for connections are inconditionally released so it doesn't come from them.

I'm noticing that if I report `option redispatch`, it still happens but much slower. So it's very possible that even regular retries leak a little bit on connection failures and that we never notice them.

Still digging...

**Comment by wtarreau** [comment]
I'm seeing that alloc_dst_address() in connect_server() blindly overwrites any previous value and is the cause for the main leak. Now with just this, it becomes as low as when there's no redispatch:
```
@@ -1270,6 +1270,7 @@ int connect_server(struct stream *s)
        srv = objt_server(s->target);
 
        if (!(s->flags & SF_ADDR_SET)) {
+               sockaddr_free(&s->si[1].dst);
                err = alloc_dst_address(&s->si[1].dst, srv, s);
                if (err != SRV_STATUS_OK)
                        return SF_ERR_INTERNAL;
```

Still trying to figure the cause for the remaining low leak.

**Comment by wtarreau** [comment]
I said bullshit, there's no other leak, the remaining one is only the local objects kept in thread caches, that's the same for connections:
```
  - Pool sockaddr (152 bytes) : 3397 allocated (516344 bytes), 3397 used (~3335 by thread caches), needed_avg 3393, 0 failures, 1 users, @0xba1e20 [SHARED]
  - Pool connection (216 bytes) : 5870 allocated (1267920 bytes), 5870 used (~5821 by thread caches), needed_avg 5866, 0 failures, 1 users, @0xb9ec70 [SHARED]
```
Thus it seems there's only the one above that was missing.

**Comment by wtarreau** [comment]
Now fixed and backported to 2.4. You can apply the attached fix to your tree if you want. Otherwise it will be in 2.4.11. I'm closing to keep the tracker clean but feel free to comment or reopen if needed. Thank you for your detailed report, that was really helpful!

[0001-BUG-MEDIUM-backend-fix-possible-sockaddr-leak-on-red.patch.txt](https://github.com/haproxy/haproxy/files/7774034/0001-BUG-MEDIUM-backend-fix-possible-sockaddr-leak-on-red.patch.txt)

**Comment by danielleontiev** [comment]
Thanks a lot for the super fast fix and explanation!
